fix(amf): Fix mobility updating registration

brunohcfaria opened the pull request:
<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

This PR fixes the Mobility Updating Registration Request reported in #1576 [PR](https://github.com/magma/magma/pull/15176). which is marked as Draft and still had problems with duplicate sessions.

In summary, the AMF was not handling the Registration Request with Mobility Updating Type. That issue was observed in laboratory with SIMCOM devices. The test case was moving out of coverage and going back while running an iperf session. It was observed that when UE restablished the RRC Connection with the gNB it was resuming the NAS signaling by sending a Registration Request with type Mobility Updating. The AMF was unable to handle the procedure correctly leading to disconnection of the user.


## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

The fix was tested with the same test-case presented above. The UE was moved out of coverage and it comes back which triggered the Mobility Updating procedure. The iperf session and pings where reestablished every time confirming the data session was kept.

![fix_mobility_updating_registration_type_screenshot](https://github.com/magma/magma/assets/33635129/cc08418c-648a-402f-b59e-d03a31ee60ed)


## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

## Security Considerations

<!--
    Comment on potential security impact. Could the change create or 
    eliminate weaknesses? STRIDE, OWASP Top 10, or RFC 3552 may help.
-->

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Mobility Update Regsitration Type handle

    Fixes a bug where the Registration Request with Mobility
    Registration Updating type was handled as a new registration,
    causing duplicate sessions in AMF and SMF.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Updated AMF test procedures for proper MobilityUpdated and Periodic
Registration

Modified the tests to not expect Identity Procedure when the TMSI
is matched in guti hashtable.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Updated AMF test procedures for proper MobilityUpdated and Periodic
Registration

Modified the tests to not expect Identity Procedure when the TMSI
is matched in guti hashtable.

Signed-off-by: Bruno Faria <bruno.faria@ektrum.com>

brunohcfaria committed:
fix(amf): Fixing lint errors

Signed-off-by: Bruno Faria <brunohcfaria@gmail.com>

brunohcfaria committed:
fix(amf): Fixing lint errors

Signed-off-by: Bruno Faria <brunohcfaria@gmail.com>

brunohcfaria committed:
fix(amf): Fixing lint errors

Signed-off-by: Bruno Faria <brunohcfaria@gmail.com>

lucasgonze reviewed (commented):
Thank you for your service. I can see this one has been an inordinate amount of work. However there are still many failing tests. As far as I can tell these are all related to Bazel. 

I'm happy to talk this one out in real time if it helps.

dependabot[bot] committed:
chore(deps): bump golang.org/x/crypto in /feg/gateway (#15639)

Bumps [golang.org/x/crypto](https://github.com/golang/crypto) from 0.0.0-20220722155217-630584e8d5aa to 0.31.0.
- [Commits](https://github.com/golang/crypto/commits/v0.31.0)

---
updated-dependencies:
- dependency-name: golang.org/x/crypto
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kmu-tm committed:
fix(amf): Correct MSIN decoding for BCD-encoded deconcealed IMSI (#15499)

- Update M5GSUCIRegistrationServiceClient.cpp for proper SUCI handling
- Modify nas_proc.cpp to fix MSIN decoding process
- Adjust test_amf_procedures.cpp for debugging

Signed-off-by: kmu-tm <kamar.mubasier@telenormaritime.com>

brunohcfaria committed:
Merge branch 'fix_mobility_updating_registration' of github.com:brunohcfaria/magma into fix_mobility_updating_registration
